### PR TITLE
[flutter_tools] Support profile and release builds on Linux

### LINKED ISSUE #38478: Implement release mode for Linux
Currently a release build (from the `flutter` tool perspective) is using a debug library build, in JIT mode; the only part that's different is anything at the Runner level (e.g., build flags for the native Runner code).

We should be building and uploading a release build of the engine library, and using that with AOT for release builds, as on mobile.

### LINKED ISSUE #39665: Implement profile mode for Linux
Blocked on #38478. Once release mode is working, there are some tweaks necessary to the native build projects, and possibly to the tooling, to finish wiring up profile.

Some of the latter (e.g., adding profile mode to the Makefile; it wouldn't do much for now, but really neither does release mode) could be done in advance.

### PR DESCRIPTION
**Pull request by MarcusTomlinson**
## Description

Tool updates to support profile and release builds on Linux.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/39665
Fixes https://github.com/flutter/flutter/issues/38478

## Tests

'ProfileBundleLinuxAssets copies artifacts to out directory'
'ReleaseBundleLinuxAssets copies artifacts to out directory'

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Nice!

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:157]: .packages isn't needed since that is an input to app.dill

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:158]: Likewise, `dart` isn't used by this target

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:189]: Same here

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:223]: app.so

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:280]: app.so

**Comment by MarcusTomlinson** [comment]
**NOTE:** For this to land, we'd need to either upload release builds of the GLFW embedder, or wait until the switch to GTK is complete. Until then, these changes will require running flutter with --local-engine.

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (MarcusTomlinson) [reply packages/flutter_tools/lib/src/build_system/targets/linux.dart:223]: Whoops :P

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (MarcusTomlinson) [reply packages/flutter_tools/lib/src/build_system/targets/linux.dart:158]: Thanks! Will fix these now.

**Comment by jonahwilliams** [comment]
This is awesome @MarcusTomlinson

**Comment by stuartmorgan** [comment]
@MarcusTomlinson I didn't do GLFW uploading since I wasn't sure where this would come in relative to switching to GTK; let me know if you want me to start uploading GLFW release and profile temporarily to unblock this. (It's a bit fiddly, and I'm already familiar with the process, so it's probably easier for me to do it if that's the route we want to go).

**Comment by MarcusTomlinson** [comment]
@stuartmorgan, yeah I think that's a good idea. Let’s get release and profile builds working with the existing GLFW embedder first. GTK can drop in later once it’s ready.

**Comment by jonahwilliams** [comment]
Blocked by https://github.com/flutter/flutter/issues/57248

**Comment by stuartmorgan** [comment]
@MarcusTomlinson Working on getting things uploaded, but presumably this is also blocked on updating the GLFW embedding to actually load the `.so`; is there a PR for that yet?

(I suggest adding a version of https://github.com/flutter/engine/blob/master/shell/platform/windows/client_wrapper/include/flutter/dart_project.h to the GFLW embedding and restructuring the other classes to use it, so that that class can handle vending the library path to the internals without the runners needing messy build-type-specific logic.)

**Comment by MarcusTomlinson** [comment]
You presume correctly @stuartmorgan :) Sorry, I somehow thought the update to GLFW was being handled elsewhere. I'll prep a PR for this.

**Comment by MarcusTomlinson** [comment]
@stuartmorgan, I've just pushed https://github.com/flutter/engine/pull/18423. It's far more simplistic than the dart_project.h approach you suggested, though I'm not sure that it really needs to be any more complex than that?

**Comment by stuartmorgan** [comment]
> @stuartmorgan, I've just pushed [flutter/engine#18423](https://github.com/flutter/engine/pull/18423). It's far more simplistic than the dart_project.h approach you suggested, though I'm not sure that it really needs to be any more complex than that?

Putting the `.so` in `flutter_assets` (which is why you don't need the project structure) seems strange to me, probably because for iOS and macOS the corresponding `.framework` contains the assets folder. But I guess in JIT mode `flutter_assets` does contain the corresponding data.

@jonahwilliams Thoughts on that structure? Is there any crisp definition of what should and shouldn't be in `flutter_assets`?

**Comment by jonahwilliams** [comment]
`flutter_assets` can contain whatever we want it to, if putting the .so there is convenient from a build & bundle perspective I'd say it's fine.

File: packages/flutter_tools/bin/tool_backend.dart
Comment (jonahwilliams) [inline packages/flutter_tools/bin/tool_backend.dart:53]: nit: put the `else if` on the same line as the closing `}`

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit.

LMK if you need help resolving the conflicts

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
I'd like to have the build target structure match what I did in Windows, both to reduce duplication, and to make it less confusing for maintaining both going forward.

(Also, I believe the consensus from the last Linux meeting was that people generally thought the libapp.so shouldn't be in the assets directory, so there will need to be an engine change. I realized there's no need to go all the way to making a DartProject class for the GLFW embedding though; you can just add a new path to the properties struct.)

**Comment by MarcusTomlinson** [comment]
> I'd like to have the build target structure match what I did in Windows

By this do you mean, put the dynamic library in the same folder as the executable? I was thinking it might make more sense in the /lib folder next to `libflutter_linux_glfw.so`.

**Comment by stuartmorgan** [comment]
> > I'd like to have the build target structure match what I did in Windows
> 
> By this do you mean, put the dynamic library in the same folder as the executable? I was thinking it might make more sense in the /lib folder next to `libflutter_linux_glfw.so`.

Sorry, I meant the `Target` subclass code structure. Specifically, re-using the existing elf AOT target instead of copying it. See `.../targets/windows.dart` in https://github.com/flutter/flutter/pull/57749

I don't have strong feelings about having it next to the executable vs. in a lib folder in the short term. Ideally I'd like everything to be in a lib/ folder since that feels more standard for unix, but I lost that when switching to CMake because the auto-bundler puts all the libs next to the executable and I didn't see how to override that behavior (short of doing all the copying of libs by hand). But if you have the make-fu to fix that, lib/ sounds great :)

**Review by jonahwilliams — COMMENTED** [review]
Do we need to bump the template version?

**Comment by stuartmorgan** [comment]
> Do we need to bump the template version?

Good call. I forgot to do that for Windows :(

**Review by jonahwilliams — APPROVED** [review]
Still LGTM!

File: packages/flutter_tools/lib/src/build_system/targets/linux.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/build_system/targets/linux.dart:194]: There's a bunch of duplicated logic across these three targets. It would be good to use inheritance and/or dependent targets to share code (see Windows, which has a mix of both, as an example; it's not necessary perfect in that I don't have a good argument for that mix vs. going entirely to inheritance or dependencies, but it does avoid repeating logic 2-3 time).

File: packages/flutter_tools/templates/app/linux.tmpl/CMakeLists.txt.tmpl
Comment (stuartmorgan) [inline packages/flutter_tools/templates/app/linux.tmpl/CMakeLists.txt.tmpl:87]: There's only one file, isn't there? I'm not sure how there could be stale files in this structure.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
One more comment on the target structure, but it's looking good.

Thanks for all the trailblazing on this! (And for resyncing with my change; I was actually going to do that so that you didn't have to clean up the fallout of me landing that one first.)

**Review by stuartmorgan-g — APPROVED** [review]
LGTM!

**Comment by stuartmorgan** [comment]
(That is, once you've fixed the test failures, but I don't expect that to require re-review.)

**Comment by jonahwilliams** [comment]
This should be good to land now. I'll mark with `waiting for tree to go green` so it can happen automagically.

**Pull request merged**
